File: app/src/main/java/com/daniel/mobilepauker2/model/notification/NotificationService.java
package com.daniel.mobilepauker2.model.notification;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import androidx.annotation.NonNull;
import androidx.core.app.JobIntentService;
import android.util.SparseLongArray;

import com.daniel.mobilepauker2.PaukerManager;
import com.daniel.mobilepauker2.model.xmlsupport.FlashCardXMLPullFeedParser;
import com.daniel.mobilepauker2.utils.Log;

import java.io.File;
import java.io.IOException;
import java.net.URI;

public class NotificationService extends JobIntentService {
    private static final int JOB_ID = 0;

    @Override
    protected void onHandleWork(@NonNull Intent intent) {
        setAlarm();
    }

    private void setAlarm() {
        Log.d("NotificationService::setAlarm", "Entered");
        AlarmManager alarmManager = (AlarmManager) getSystemService(Context.ALARM_SERVICE);
        Intent alarmIntent;
        PendingIntent pendingIntent;

        long newestTime = getNewestExpireTime();
        Log.d("NotificationService::setAlarm", "Newest Time: " + newestTime + ". Now is: " + System.currentTimeMillis());

        if (newestTime > -1 && alarmManager != null) {
            alarmIntent = new Intent(this, AlarmNotificationReceiver.class);
            pendingIntent = PendingIntent.getBroadcast(this, 0, alarmIntent, 0);
            alarmManager.set(AlarmManager.RTC_WAKEUP, newestTime, pendingIntent);
            Log.d("NotificationService::setAlarm", "Alarm set");
        }
    }

    private long getNewestExpireTime() {
        PaukerManager paukerManager = PaukerManager.instance();
        File[] files;
        try {
            files = paukerManager.listFiles(this);
        } catch (SecurityException e) {
            files = new File[0];
        }

        long newestTime = -1;
        URI uri;
        FlashCardXMLPullFeedParser parser;

        for (File file : files) {
            if (!(paukerManager.isSaveRequired() && paukerManager.getCurrentFileName().equals(file.getName()))) {
                try {
                    uri = paukerManager.getFilePath(this, file.getName()).toURI();
                    parser = new FlashCardXMLPullFeedParser(uri.toURL());
                    SparseLongArray map = parser.getNextExpireDate();
                    if (map.get(0) > Long.MIN_VALUE) {
                        if (map.get(1, 0) > 0) {
                            return 0;
                        } else {
                            if (newestTime == -1 || map.get(0) < newestTime) {
                                newestTime = map.get(0);
                            }
                        }
                    }
                } catch (IOException ignored) {
                    Log.d("NotificationService::setAlarm", "Cannot read File");
                }
            }
        }

        return newestTime;
    }

    public static void enqueueWork(Context context) {
        enqueueWork(context, NotificationService.class, JOB_ID, new Intent());
    }
}

File: app/src/main/java/com/daniel/mobilepauker2/model/notification/NotificationService.kt
package com.daniel.mobilepauker2.model.notification

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import androidx.core.app.JobIntentService
import com.daniel.mobilepauker2.PaukerManager
import com.daniel.mobilepauker2.model.xmlsupport.FlashCardXMLPullFeedParser
import com.daniel.mobilepauker2.utils.Log
import java.io.File
import java.io.IOException
import java.net.URI

class NotificationService : JobIntentService() {
    override fun onHandleWork(intent: Intent) {
        setAlarm()
    }

    private fun setAlarm() {
        Log.d("NotificationService::setAlarm", "Entered")
        val alarmManager =
            getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val alarmIntent: Intent
        val pendingIntent: PendingIntent
        val newestTime = newestExpireTime
        Log.d(
            "NotificationService::setAlarm",
            "Newest Time: " + newestTime + ". Now is: " + System.currentTimeMillis()
        )
        if (newestTime > -1 && alarmManager != null) {
            alarmIntent = Intent(this, AlarmNotificationReceiver::class.java)
            pendingIntent = PendingIntent.getBroadcast(this, 0, alarmIntent, 0)
            alarmManager[AlarmManager.RTC_WAKEUP, newestTime] = pendingIntent
            Log.d("NotificationService::setAlarm", "Alarm set")
        }
    }

    private val newestExpireTime: Long
        private get() {
            val paukerManager: PaukerManager = PaukerManager.Companion.instance()
            val files: Array<File?>?
            files = try {
                paukerManager.listFiles(this)
            } catch (e: SecurityException) {
                arrayOfNulls(0)
            }
            var newestTime: Long = -1
            var uri: URI
            var parser: FlashCardXMLPullFeedParser
            for (file in files!!) {
                if (!(paukerManager.isSaveRequired && paukerManager.currentFileName == file!!.name)) {
                    try {
                        uri = paukerManager.getFilePath(this, file!!.name).toURI()
                        parser = FlashCardXMLPullFeedParser(uri.toURL())
                        val map = parser.nextExpireDate
                        if (map!![0] > Long.MIN_VALUE) {
                            if (map[1, 0] > 0) {
                                return 0
                            } else {
                                if (newestTime == -1L || map[0] < newestTime) {
                                    newestTime = map[0]
                                }
                            }
                        }
                    } catch (ignored: IOException) {
                        Log.d(
                            "NotificationService::setAlarm",
                            "Cannot read File"
                        )
                    }
                }
            }
            return newestTime
        }

    companion object {
        private const val JOB_ID = 0
        fun enqueueWork(context: Context?) {
            enqueueWork(
                context!!,
                NotificationService::class.java,
                JOB_ID,
                Intent()
            )
        }
    }
}